src/fu-offline.c

```diff
@@ -16,51 +16,102 @@
 #include "fu-plugin-private.h"
 #include "fu-util-common.h"
 
+typedef enum {
+	FU_OFFLINE_FLAG_NONE		= 0,
+	FU_OFFLINE_FLAG_ENABLE		= 1 << 0,
+	FU_OFFLINE_FLAG_USE_PROGRESS	= 1 << 1,
+} FuOfflineFlag;
+
 struct FuUtilPrivate {
 	gchar		*splash_cmd;
 	GTimer		*splash_timer;
+	FuOfflineFlag	 splash_flags;
 };
 
 static gboolean
 fu_offline_set_splash_progress (FuUtilPrivate *priv, guint percentage, GError **error)
 {
-	g_autofree gchar *msg = NULL;
-	g_autofree gchar *str = NULL;
-	g_autoptr(GSubprocess) subprocess = NULL;
+	g_autofree gchar *str = g_strdup_printf ("%u", percentage);
+	const gchar *argv[] = { priv->splash_cmd, "system-update", "--progress", str, NULL };
+
+	/* call into plymouth if installed */
+	if (priv->splash_flags == FU_OFFLINE_FLAG_NONE) {
+		/* TRANSLATORS: console message when not using plymouth */
+		g_printerr ("%s: %u%%\n", _("Percentage complete"), percentage);
+		return TRUE;
+	}
+
+	/* fall back to really old mode that should be supported by anything */
+	if ((priv->splash_flags & FU_OFFLINE_FLAG_USE_PROGRESS) == 0) {
+		argv[1] = "display-message";
+		argv[2] = "--text";
+	}
+	return fu_common_spawn_sync (argv, NULL, NULL, 200, NULL, error);
+}
+
+static gboolean
+fu_offline_set_splash_mode (FuUtilPrivate *priv, GError **error)
```

Comment:
I feel like the logic in this function is a little backwards compared to the rest of the codebase.  Rather than return TRUE in final function pass scenario all the fallbacks come down to FALSE.

It seems like it should be:

```
if (!ret) {
	g_autoptr(GError) error_local = NULL;
	/* fancy mode failed, fallback */
	/* do stuff */
	if (!ret) {
		/* shutdown failed */
		return FALSE;
	}
return TRUE;
```

Reply:
Agreed, with switch around.

Reply:
Fixed.